waspc/src/Analyzer/Type.hs

```diff
@@ -0,0 +1,29 @@
+module Analyzer.Type
+  ( Type (..),
+    DictEntryType (..),
+    dictEntryRequired,
+  )
+where
+
+-- | All possible types in Wasp.
+data Type
+  = DeclType String
+  | EnumType String
+  | DictType [DictEntryType]
+  | ListType Type
+  | StringType
+  | NumberType
+  | ExtImportType
+  | JSONType
+  | PSLType
+
+-- | The type of an entry in a `Dict`.
+data DictEntryType
```

Comment:
I believe it will make the type checker code clearer if pattern matching is done on `DictEntry` and `DictOptionalEntry` vs. `DictEntry name typ True` and `DictEntry name typ False`.